integrations/nifi/nifi-rest/nifi-rest-client/nifi-rest-client-api/src/main/java/com/thinkbiganalytics/nifi/rest/client/AbstractNiFiControllerServicesRestClient.java

```diff
@@ -81,6 +95,161 @@ public ControllerServiceDTO updateStateById(@Nonnull final String id, @Nonnull f
     }
 
     /**
+     * Updates a controller service
+     * This will first disable the service, stop all referencing components, enable the service, reset the state of the components back to their prior state
+     *
+     * @param controllerService the service to update with the updated properties
+     * @return the updates service
+     */
+    @Override
+    public ControllerServiceDTO updateServiceAndReferencingComponents(ControllerServiceDTO controllerService) {
```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Refactor this method to reduce its Cognitive Complexity from 31 to the 15 allowed. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS3776)